scipy/stats/_continuous_distns.py

```diff
@@ -906,8 +906,43 @@ def _cdf(self, x, a, b):
             f2=lambda x_, a_, b_: beta._cdf(x_/(1+x_), a_, b_))
 
     def _ppf(self, p, a, b):
-        r = sc.betaincinv(a, b, p)
-        return r / (1 - r)
+        # _ppf_direct is the direct way to compute the ppf (except for the
+        # additional safety mechanism described in the comment below)
+        # f2 is based on the inversion of the expression using beta._sf in
+        # self.cdf, see comments there.
+        # note that p=1 has to be handled separately to avoid division by 0
+
+        def _ppf_direct(p_, a_, b_):
+            r = stats.beta.ppf(p_, a_, b_)
+            # note: even if we use _ppf_direct only for q < 0.5,
+            # there is no guarantee that r will not be close to 1. if r is
+            # too close to 1, r/(1-r) can cause numerical issues
+            # for fixed p and a, by making b very small, one can get r
+            # arbitrarily close to 1. Example:
+            # stats.beta.ppf(0.1, 0.5, 0.001) == 1.0
+            # so first compute r and then fall back to the "indirect approach"
+            # (f2) if r is too close to 1
+            out = _lazywhere(
+                r < 0.9999,
+                [r, p_, a_, b_],
+                lambda r1, p1, a1, b1: r1 / (1 - r1),
+                f2=lambda r1, p1, a1, b1: 1/stats.beta.isf(p1, b1, a1) - 1
+            )
+            return out
+
+        # needed to avoid TypeError when casting float64 to float32
+        # (check_ppf_dtype in tests/test_continuous_basic.py)
+        p = np.asarray(p, dtype=float)
+        out = _lazywhere(
+            p < 0.5, [p, a, b],
+            _ppf_direct,
+            f2=lambda p_, a_, b_: _lazywhere(
+                p_ < 1,
+                [p_, a_, b_],
+                lambda p1, a1, b1: 1/stats.beta.isf(p1, b1, a1) - 1,
+                fillvalue=np.inf)
+        )
+        return out
```

Comment:
The last `_lazywhere` with the `fillvalue` doesn't appear to be necessary, because `_ppf` is not passed `p > 1` by the distribution infrastructure.

Besides that, it looks like there are only two approaches, "direct" and using `beta.isf`. Ultimately, the condition to choose between them is `r < 0.9999`. It sounds like `r` _can_ be big when `p < 0.5`, but `r` is never big for `p > 0.5`, hence the purpose of one `_lazywhere` is to branch for `p < 0.5` so that you only need to calculate `r` in the `p < 0.5` case. 

There are some cases in which this is beneficial, and there are some for which the overhead slows things down overall (never mind the complexity). Do you have a guess for how often `r > 0.9999` (and the nested `_lazywhere` has some chance of being beneficial?)

My guess is that it will be rare for the `_lazywhere` to provide a performance boost, and the complexity can be avoided by:
* Calculate _all_ the `r`s. 
* One `_lazywhere` to branch for `r < 0.9999`.

Or what I would prefer:
```suggestion
        r = stats.beta._ppf(p, a, b)
        out =  r / (1 - r)
        out[r < 0.9999] = 1/stats.beta._isf(p, b, a) - 1
```

Reply:
Yeah, good point. I think it is already much simpler now after getting rid of the extra lazywhere to handle the case p=1.0. My idea was indeed to prevent evaluating a special function twice, it is hard to say what the use cases are. If you prefer, we can add your proposed version.

Reply:
I'd prefer the simplified version. `betaprime` has been in SciPy since the beginning, and the slow speed of the default ppf implementation was never reported as an issue. gh-17562 offered a huge (unrequested) speed boost, so all the users are going to see either way is that PPF got a lot faster. I don't think the extra complexity here is warranted given that it's slower in all but the exceptional (many `r > 0.9999`) case.

Reply:
I agree and I changed the code accordingly